tools/src/bin/output-file.rs

```diff
@@ -199,6 +199,98 @@ fn read_detailed_file_info(path: &str, index_path: &str) -> Option<json::Object>
+/// The interesting questions are:
+/// - What to do the start and end of the file.
+/// - What to do when the uncovered gap is between a hit and a miss.  Extra
+///   information about the AST or nesting contexts might help
+///
+/// Our arbitrary decisions here are:
+/// - Leave the starts and ends of file uncovered.  This is more realistic but
+///   at the cost of this realism making it less obvious that interpolation is
+///   present in the rest of the file, potentially leading to bad inferences.
+///   - We attempt to mitigate this by making sure the hover information makes
+///     it clear when interpolation is at play so if someone looks into what's
+///     going on they at least aren't misled for too long.
+/// - Maximally interpolate hits over misses.  Our goal is that people's eyes
+///   are drawn to misses.  This interpolation strategy makes sure that the
+///   start and end of a run of misses are lines that are explicitly detected
+///   as misses.
+///
+fn interpolate_coverage(mut raw: Vec<i32>) -> Vec<i32> {
+    // We don't interpolate at the start or end of files, so start with already
+    // having a valid -1 interpolation value.
+    let mut have_interp_val = true;
+    let mut interp_val = -1;
+    // This value will never be used because we set have_interp_val to true
+    // above which means we won't calculate an interpretation with this value.
+    let mut last_noninterp_val = -1;
+    for i in 0..raw.len() {
+        let val = raw[i];
+        // If we have a valid coverage value (=0 is miss, >0 is hit) then leave
+        // the value as is, remember this value for interpolation and note that
+        // we'll need to compute our next interpolation value.
+        if val >= 0 {
+            last_noninterp_val = val;
+            have_interp_val = false;
+            continue;
+        }
+        // Not a valid value, so we need to interpolate.
+
+        // Did we already calculate our interpolation value?  If so, keep using
+        // it.  (Note that at the start of the file we start our overwriting -1
+        // with -1.)
+        if have_interp_val {
+            raw[i] = interp_val;
+            continue;
+        }
+
+        // Check the next lines until we find a value that's >= 0.  If we don't
+        // find any, then our end-of-file logic wants us to maintain a -1, so
+        // configure for that base-case.
+        have_interp_val = true;
+        interp_val = -1;
+        for j in (i + 1)..raw.len() {
+            let next_val = raw[j];
+            if next_val == -1 {
+                continue;
+            }
+            // We've found a value which means that both last_noninterp_val and
+            // next_val are >= 0.  (last_noninterp_val can never be -1 because
+            // we start the loop with have_interp_val=true.)
+
+            // Favor hits over misses (see func doc block for rationale).
+            if val > 0 || last_noninterp_val > 0 {
```

Comment:
Shouldn't this be `next_val > 0` instead of `val > 0`? Tangentially, this function looks like a nice candidate for a few unit tests.

Reply:
Ah, yes, you're absolutely right on both counts.  When I implemented this and it seemed to work immediately I didn't ask too many questions and I think I got fixated on other things that were more obviously on fire, but I absolutely should have implemented an in-file test func.  Thanks very much for calling this out!

Your specific example is now caught by the:
```rust
        vec![
            vec![1, -1, 0, -1, 1],
            vec![1, -2, 0, -2, 1]
        ],
```
case.  With the failure having been reported as:
```
---- test_interpolate_coverage stdout ----
thread 'test_interpolate_coverage' panicked at 'assertion failed: `(left == right)`
  left: `[1, -2, 0, -3, 1]`,
 right: `[1, -2, 0, -2, 1]`', src/bin/output-file.rs:339:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The following test case also uncovered the lack of a `break` at the bottom of the loop!
```rust
        // combine all of the above (except for whole file), single interp each.
        vec![
            vec![-1, -1, 0, -1, 0, -1, 1, -1, 1, -1, 1, -1, 0, -1],
            vec![-1, -1, 0, -3, 0, -2, 1, -2, 1, -2, 1, -2, 0, -1]
        ],
```

Manifesting the failure:
```
---- test_interpolate_coverage stdout ----
thread 'test_interpolate_coverage' panicked at 'assertion failed: `(left == right)`
  left: `[-1, -1, 0, -3, 0, -3, 1, -2, 1, -2, 1, -2, 0, -1]`,
 right: `[-1, -1, 0, -3, 0, -2, 1, -2, 1, -2, 1, -2, 0, -1]`', src/bin/output-file.rs:339:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```